text/0019-stricter-validation.md

```diff
@@ -0,0 +1,77 @@
+
+In sum: To see why the second case is acceptable, we have to understand the interaction between permissive mode and strict mode. This is subtle and difficult to explain. Instead, we want to implement strict mode as a standalone feature, which does not depend on features only present in permissive mode.
+
+## Detailed design
+
+We propose to change strict-mode validation so that it has the following features taken from permissive mode:
+* singleton boolean `True`/`False` types (which have 1/ special consideration of them with `&&`, `||`, and conditionals, and 2/ subtyping between them and `Boolean`, as is done now for permissive mode).
+* depth subtyping, but not width subtyping, to support examples like `{ foo: True } <: { foo: bool }`.
+* a requirement that arguments to `==` have the same type _unless_ we know that comparison will always be false, in which case we can give the `==` type `False`. The same goes for `contains`, `containsAll`, and other constructs that use `==` under the covers.
+* a requirement that branches of a conditional have the same type _unless_ we can simplify the conditional due to the guard having a `True`/`False` type.
+* (no union types)
+
+The current way we implement strict mode will change to accommodate these additions. We propose the following:
+
+1. Change the Rust code to implement strict mode by reusing the logic of permissive mode​ and using a flag to perform more restrictive checks. For example, when given `User` and `Admin` entity types, the code performing the least-upper-bound computation will return union type `User|Admin` in permissive mode, and will fail in strict mode. And, the code for conditionals would force branches' types to match in strict mode, but not in permissive mode. Etc.
+2. When checking in strict mode, we construct a new AST that performs the needed transformations, as in step 2 today. 
+3. Either way, we can add type annotations. For strict mode, they could be added during the transformation; for permissive mode, no transformation is needed. No checking is done after annotations are added, since the checking is always done, regardless of mode, prior to the transformation.
+
+Permissive mode validation considers different `Action`s to have distinct types, e.g., `Action::"view"` and `Action::"update"` don't have the same type. This is in anticipation supporting distinct actions having distinct attribute sets. Today, the least upper bound of any two actions is the type _AnyEntity_, which is a general supertype of all entities, used internally. Such a type supports expressions that are found in the policy scope such as `action in [Action::"view", Action::"edit"]`. This expression typechecks because the least upper bound of `Action::"view"` and `Action::"edit"` is _AnyEntity_ so the expression is type `Set<`_AnyEntity_`>`. Therefore in our amended strict mode we either need to:
+1. Support _AnyEntity_ as part of depth subtyping.
```

Comment:
Regarding `AnyEntity`: It's also used to support typechecking template slots. That won't work with this definition of strict typechecking (it doesn't work with the old definition either, but we added special case to allow it). How do we want to handle them now? In a prior RFC we seemed OK with only doing template validation after linking.

Reply:
We use the `principalTypes` and `resourceTypes` information from the cross-product. In particular:

We know for a particular action what the principal and resource type must be. The slot is either on the RHS of a `==` or a `in`. If the former, you just type the slot with the principal/resource type. If the latter, you look up the possible `memberOfTypes` and use these, systematically.

Reply:
+1 for using the cross-product approach to typecheck the templates statically, independently of linking.  This is needed in order to analyze template behavior wrt all possible instantiations.

Reply:
Not based on scope conditions, based on the schema. If your policy has 
```
principal == ?principal
```
in the scope, then you can give `?principal` the same type as `principal` used for the cross product. So if the current `principal` type is `User`, typecheck the policy with `?principal:User` too. If instead we have
```
principal in ?principal
```
in the scope, then you can further extend the cross product to type `?principal` with each of the `principal` type's `memberOfTypes`. For example, if `principal` is being typed as `User` and the `memberOfTypes` for `User` in the schema are `Team` and `Region`, then you typecheck the policy first with `principal:User` and `?principal:Team` and then `principal:User` and `?principal:Region`. 

Do likewise for `?resource`. Make sense?

Reply:
The procedure you described seems to be based on the scope conditions.

Reply:
If we let template slots appear in the body of a policy, then this starts to make sense.

Reply:
This makes sense if you think of actions as (overloaded) functions with signatures.  The suggestion here is to typecheck the template against all possible action signatures.  This isn't circular---it depends just on the schema.

A naive implementation is to iterate through every single action in the schema, and typecheck the template against all of its overloaded signatures.

We can make this more efficient by looking at scope constraints, but this is an optimization, not something inherent to the algorithm that @mwhicks1 proposed.

Reply:
@emina described it well. Focusing on the shape of the current policy is an optimization. You could ignore it, but you'd end up with repeated effort.

Reply:
Proof is in the pudding of course: We should be able to prove validation soundness of this approach!